modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java

```diff
@@ -84,50 +80,115 @@ public class SchemaManager extends Producer<SchemaEvent, SchemaEventParameters>
+            } else {
+                serialized = schemas.get(INITIAL_SCHEMA_VERSION);
+
+                assert serialized != null;
             }
-        });
+        }
+
+        tablesCfg.tables().any().columns().listen(this::onSchemaChange);
     }
 
     /**
      * Listener of schema configuration changes.
      *
-     * @param schemasCtx Schemas configuration context.
+     * @param ctx Configuration context.
      * @return A future.
      */
-    private CompletableFuture<?> onSchemaCreate(ConfigurationNotificationEvent<SchemaView> schemasCtx) {
+    private CompletableFuture<?> onSchemaChange(ConfigurationNotificationEvent<NamedListView<ColumnView>> ctx) {
         if (!busyLock.enterBusy()) {
             return failedFuture(new IgniteInternalException(NODE_STOPPING_ERR, new NodeStoppingException()));
         }
 
         try {
-            long causalityToken = schemasCtx.storageRevision();
+            ExtendedTableView tblCfg = (ExtendedTableView) ctx.config(ExtendedTableConfiguration.class).value();
 
-            ExtendedTableConfiguration tblCfg = schemasCtx.config(ExtendedTableConfiguration.class);
+            int verFromUpdate = tblCfg.schemaId();
 
-            UUID tblId = tblCfg.id().value();
+            UUID tblId = tblCfg.id();
+
+            String tableName = tblCfg.name();
+
+            SchemaDescriptor schemaDescFromUpdate = SchemaUtils.prepareSchemaDescriptor(verFromUpdate, tblCfg);
+
+            if (searchSchemaByVersion(tblId, schemaDescFromUpdate.version()) != null) {
+                return completedFuture(null);
+            }
+
+            byte[] curSchemaDesc = schemaById(tblId, verFromUpdate);
+
+            if (verFromUpdate != INITIAL_SCHEMA_VERSION) {
+                SchemaDescriptor oldSchema = searchSchemaByVersion(tblId, verFromUpdate - 1);
+                assert oldSchema != null;
 
-            String tableName = tblCfg.name().value();
+                NamedListView<ColumnView> oldCols = ctx.oldValue();
+                NamedListView<ColumnView> newCols = ctx.newValue();
 
-            SchemaDescriptor schemaDescriptor = SchemaSerializerImpl.INSTANCE.deserialize((schemasCtx.newValue().schema()));
+                schemaDescFromUpdate.columnMapping(SchemaUtils.columnMapper(
+                        oldSchema,
+                        oldCols,
+                        schemaDescFromUpdate,
+                        newCols));
+            }
+
+            long causalityToken = ctx.storageRevision();
+
+            CompletableFuture<?> createSchemaFut = createSchema(causalityToken, tblId, tableName, schemaDescFromUpdate);
+
+            if (curSchemaDesc == null) {
+                try {
+                    byte[] serialized = SchemaSerializerImpl.INSTANCE.serialize(schemaDescFromUpdate);
 
-            CompletableFuture<?> createSchemaFut = createSchema(causalityToken, tblId, tableName, schemaDescriptor);
+                    createSchemaFut.thenCompose(t -> metastorageMgr.put(
```

Comment:
you need to CAS here instead of `put`

Reply:
listener is called on each node, i have some questions :
1. there is no cas at all, do you mean getAndPut with further assertion ? 
2. seems we can`t use cas here.